etcdserver/snapshot_hub.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+type snapshot struct {
+	// storing the raft snapshot. the data in raftsnap is v2 store.
+	raftsnap raftpb.Snapshot
+	// rc is a reader that provides the data of KV snapshot.
+	rc io.ReadCloser
+	// dataSize specifies the size of snapshot data.
+	dataSize int64
+}
+
+type snapshotHub struct {
+	// dir to save snapshot data
+	dir string
+
+	// send empty to reqsnapc to notify the channel receiver to send back latest
+	// snapshot to snapc
+	reqsnapc chan struct{}
+	// a chan to receive the requested snapshot
+	// hub will receive from the chan immediately after it sends empty to reqsnapc
+	snapc chan snapshot
+	// host received snapshots
+	snaps []snapshot
+}
+
+func newSnapshotHub(dir string) *snapshotHub {
+	return &snapshotHub{
+		dir:      dir,
+		reqsnapc: make(chan struct{}),
+		snapc:    make(chan snapshot),
+	}
+}
+
+// CreateSnapshot creates snapshot with latest index. The returned snapshot
+// is the mark of snapshot, and caller could call SnapshotData to get its data.
+// TODO: the function should be unblock and finish in 1ms. It may use ~20ms
+// in the peak for 2 million keys.
+func (sh *snapshotHub) CreateSnapshot() raftpb.Snapshot {
```

Comment:
This function does not actually create the snapshot. It sends out a snapshot request. Another routine receives the request and creates the snapshot. Basically, this function only cares about the logic to get the snapshot. It does not care if the snapshot is a created one or we need to create a new one.